### PR TITLE
fix: improve wallets ui

### PR DESCRIPTION
**Pull request by lelenei**
fix: #2117
fix: #2261

### CODE REVIEW

**Comment by Jack-Works** [comment]
about #2261, we have [concurrent mode](https://reactjs.org/docs/concurrent-mode-intro.html) enabled. So you don't need to check the "loading" state manually. An alternative way is: trigger the suspended state and let the `<Suspense>` fallback show the loading state.

**Comment by zhouhanseng** [comment]
Here's an example of Integrating with suspense: https://github.com/DimensionDev/Maskbook/pull/2221

File: packages/maskbook/src/web3/hooks/useAssetsDetailedCallback.ts [hunk 1]
```diff
@@ -27,6 +27,7 @@ export function useAssetsDetailedCallback(tokens: (EtherTokenDetailed | ERC20Tok
 
     // filter out tokens in blacklist
     return {
+        loading: loading && ethloading,
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/hooks/useAssetsDetailedCallback.ts:30]:
```suggestion
        loading: loading && loadingEtherToken,
```

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx [hunk 2]
```diff
@@ -73,15 +76,30 @@ export interface WalletAssetsTableProps extends withClasses<KeysInferFromUseStyl
...
+        setViewLength(
+            detailedTokens.length > MAX_TOKENS_LENGTH
+                ? haveETH
+                    ? MAX_TOKENS_LENGTH - 1
+                    : MAX_TOKENS_LENGTH
+                : detailedTokens.length,
+        )
+    }, [detailedTokens.length, haveETH])
+
+    if (!chainIdValid) return null
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx:102]:
```suggestion
```

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx [hunk 3]
```diff
@@ -113,6 +124,9 @@ export const WalletContent = forwardRef<HTMLDivElement, WalletContentProps>(func
     }, [])
     //#endregion
 
+    //#collectibles loading
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:127]: Maybe move `useERC721TokenDetailed` to this component, life will be simpler.

Comment (nuanyang233) [reply packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:127]: The best way is trigger suspense fallback loading..so we have a library to turn the async function to suspend? like 
 `getERC721TokenDetailed`

Comment (guanbinrui) [reply packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:127]: Sounds great! Let's rewrite those utility hooks to suspends.

Comment (lelenei) [reply packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:127]: ok

Comment (zhouhanseng) [reply packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:127]: Since `useERC721TokenDetailed` uses other hooks, it still need to be named `useERC721TokenDetailed`.

Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:127]: Use `#region` and `#endregion`.

File: packages/maskbook/src/web3/hooks/useERC721TokenDetailed.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@ import { EthereumTokenType, ERC721TokenDetailed } from '../types'
 import { useChainId } from './useChainState'
 import { formatChecksumAddress } from '../../plugins/Wallet/formatter'
 import { useERC721TokenContract } from '../contracts/useERC721TokenContract'
-import { useSingleContractMultipleData } from './useMulticall'
+import { MulticalStateType, useSingleContractMultipleData } from './useMulticall'
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/hooks/useERC721TokenDetailed.ts:8]: Unused import?

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+let token_: ERC721TokenDetailed | undefined = undefined
+async function ERC721TokenDetailed_(address: string, token?: Partial<ERC721TokenDetailed>) {
+    const chainId = useChainId()
+    const erc721TokenContract = useERC721TokenContract(address)
+
+    // compose calls
+    const { names, callDatas } = useMemo(
+        () => ({
+            names: ['name', 'symbol', 'baseURI'] as 'name'[],
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:16]: this type cast looks strange

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx [hunk 6]
```diff
@@ -76,16 +79,28 @@ export interface WalletAssetsTableProps extends withClasses<KeysInferFromUseStyl
...
 
-    if (!detailedTokens.length) return null
+    const chainIdValid = useChainIdValid()
+
+    const haveETH = detailedTokens.some((x) => isETH(x.token.address))
+    const [viewLength, setViewLength] = useState(haveETH ? MAX_TOKENS_LENGTH - 1 : MAX_TOKENS_LENGTH)
+
+    useEffect(() => {
+        setViewLength(
+            detailedTokens.length > MAX_TOKENS_LENGTH
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx:97]: It looks like a computed value, please don't sync it with an effect, compute it directly (or if you have special reason please comment it)

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx [hunk 7]
```diff
@@ -157,42 +172,63 @@ export function WalletAssetsTable(props: WalletAssetsTableProps) {
...
         </div>
     )
 
+    const filterETH = (a: AssetDetailed) => isETH(a.token.address)
+    const excludeETH = (a: AssetDetailed) => !isETH(a.token.address)
+    const excludeZero = (a: AssetDetailed) => new BigNumber(a.value?.[CurrencyType.USD] || '0').isGreaterThan(0)
+    const filter = (a: AssetDetailed) =>
+        Number(price) !== 0 ? new BigNumber(a.value?.[CurrencyType.USD] || '0').isGreaterThan(price) : true
+    const sort = (a: AssetDetailed, b: AssetDetailed): number =>
+        new BigNumber(b.value?.[CurrencyType.USD] ?? 0).minus(a.value?.[CurrencyType.USD] ?? 0).toNumber()
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx:190]: hoist those 5 functions to the top level if possible

File: .vscode/settings.json [hunk 8]
```diff
@@ -19,5 +19,6 @@
             "url": "/.vscode/schema/locale.json"
         }
     ],
-    "cSpell.words": ["Reificated"]
+    "cSpell.words": ["Reificated"],
+    "i18n-ally.localesPaths": "packages/maskbook/src/plugins/Trader/messages"
```
Comment (Jack-Works) [inline .vscode/settings.json:23]: revert this change plz

File: packages/maskbook/src/web3/hooks/useERC721TokenDetailed.ts [hunk 9]
```diff
@@ -12,13 +12,8 @@ export function useERC721TokenDetailed(address: string, token?: Partial<ERC721To
...
 
     // compose calls
-    const { names, callDatas } = useMemo(
-        () => ({
-            names: ['name', 'symbol', 'baseURI'] as 'name'[],
-            callDatas: new Array(3).fill([]),
-        }),
-        [],
-    )
+    const names = ['name', 'symbol', 'baseURI'] as 'name'[]
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/hooks/useERC721TokenDetailed.ts:15]: why cast it as 'name'[]?

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 5 above)
Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:16]: why cast it as 'name'[]?

File: .vscode/settings.json (hunk 8 above)
Comment (lelenei) [reply .vscode/settings.json:23]: This is automatically modified by vscode

Comment (lelenei) [reply .vscode/settings.json]: this is modified by prettier

Comment (Jack-Works) [reply .vscode/settings.json]: no it's not, it's added by i18n-ally vscode plugin

File: .vscode/settings.json [hunk 10]
```diff
@@ -19,5 +19,10 @@
             "url": "/.vscode/schema/locale.json"
         }
     ],
-    "cSpell.words": ["Reificated"]
+    "cSpell.words": ["Reificated"],
+    "i18n-ally.localesPaths": [
+        "packages/dashboard/src/locales",
+        "packages/theme/src/locales",
+        "packages/maskbook/src/plugins/Trader/messages"
```
Comment (Jack-Works) [inline .vscode/settings.json:26]: 🤔or maybe you should add all locale paths. but I doubt if it works with monorepo well.

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx [hunk 11]
```diff
@@ -73,19 +76,31 @@ export interface WalletAssetsTableProps extends withClasses<KeysInferFromUseStyl
...
 
     const classes = useStylesExtends(useStyles(), props)
     const LABELS = [t('wallet_assets'), t('wallet_price'), t('wallet_balance'), t('wallet_value'), ''] as const
 
-    const [viewLength, setViewLength] = useState(MAX_TOKENS_LENGTH)
     const [more, setMore] = useState(false)
     const [price, setPrice] = useState(MIN_VALUE)
 
-    if (!detailedTokens.length) return null
+    const haveETH = detailedTokens.some((x) => isETH(x.token.address))
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx:96]:
```suggestion
    const haveETH = detailedTokens.some(filterETH)
```

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 5 above)
Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:16]: request a change here

File: .vscode/settings.json (hunk 10 above)
Comment (lelenei) [reply .vscode/settings.json:26]: I upgrade i18n plugin in vscode

File: .vscode/settings.json (hunk 8 above)
Comment (lelenei) [reply .vscode/settings.json]: Yes!

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        name: name ?? token?.name ?? '',
+        symbol: symbol ?? token?.symbol ?? '',
+        baseURI: baseURI ?? token?.baseURI ?? '',
+    } as ERC721TokenDetailed
+
+    return
+}
+
+export function getERC721TokenDetailed(address: string, token?: Partial<ERC721TokenDetailed>) {
+    if (!token_) throw ERC721TokenDetailed_(address)
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:36]: Suppose we are using this function in different places on the same page. Will this function always return the result of the first call? Cause you are using a local variable to store the fetched token.

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx [hunk 13]
```diff
@@ -157,42 +172,58 @@ export function WalletAssetsTable(props: WalletAssetsTableProps) {
...
                 setPrice(more ? MIN_VALUE : 0)
             }}>
-            <IconButton>{more ? <ExpandLessIcon /> : <ExpandMoreIcon />}</IconButton>
+            {detailedTokens.length > MAX_TOKENS_LENGTH ? (
+                <IconButton>{more ? <ExpandLessIcon /> : <ExpandMoreIcon />}</IconButton>
+            ) : null}
         </div>
     )
 
+    const filter = (a: AssetDetailed) =>
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx:184]: Please use a proper name instead.

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 12 above)
Comment (lelenei) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:36]: Token is unchanged and can be used directly in other places

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import { formatChecksumAddress } from '../../plugins/Wallet/formatter'
+import { useERC721TokenContract } from '../contracts/useERC721TokenContract'
+import { useChainId } from '../hooks/useChainState'
+import { useSingleContractMultipleData } from '../hooks/useMulticall'
+import { ERC721TokenDetailed, EthereumTokenType } from '../types'
+
+const cache = new Map<string, ERC721TokenDetailed | undefined>()
+async function ERC721TokenDetailed_(address: string, token?: Partial<ERC721TokenDetailed>) {
+    const chainId = useChainId()
+    const erc721TokenContract = useERC721TokenContract(address)
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:11]: IMO this is not correct. You're violating React hooks rules. Because this function is conditionally called. If you're intended to do this, please call those hooks in `getERC721TokenDetailed` unconditionally and rename it to `useERC721TokenDetailed`.

**Comment by guanbinrui** [comment]
Hi @lelenei Could you please paste some screenshots here?

**Comment by lelenei** [comment]
image

File: packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx [hunk 15]
```diff
@@ -187,9 +198,17 @@ export const WalletContent = forwardRef<HTMLDivElement, WalletContentProps>(func
...
-                {Flags.COTM_enabled && tabIndex === 1 ? <COTM_TokenAlbum /> : null}
-                {Flags.election2020_enabled && tabIndex === 1 ? <ElectionTokenAlbum /> : null}
+                ) : (
+                    <Suspense
+                        fallback={
+                            <Typography variant="body1" color="textSecondary">
+                                {t('loading')}
+                            </Typography>
+                        }>
+                        {Flags.COTM_enabled ? <COTM_TokenAlbum /> : null}
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardComponents/WalletContent.tsx:208]: Each plugin should have it's own Suspense

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    chainId: ChainId,
+    contract: Contract,
+    address: string,
+    token?: Partial<ERC721TokenDetailed>,
+) {
+    // compose calls
+    const names = ['name', 'symbol', 'baseURI'] as (keyof ERC721['methods'])[]
+    const callDatas = new Array(3).fill([])
+
+    const [results, calls, _, callback] = useSingleContractMultipleData(contract, names, callDatas)
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:20]: No hooks in this function, please

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    const token_ = {
+        type: EthereumTokenType.ERC721,
+        address: formatChecksumAddress(address),
+        chainId,
+        name: name ?? token?.name ?? '',
+        symbol: symbol ?? token?.symbol ?? '',
+        baseURI: baseURI ?? token?.baseURI ?? '',
+    } as ERC721TokenDetailed
+    cache.set(address, token_)
+    return
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:35]: Remove this line

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+export function useAsyncERC721TokenDetailed(address: string, token?: Partial<ERC721TokenDetailed>) {
+    const chainId = useChainId()
+    const erc721TokenContract = useERC721TokenContract(address) as Contract
+    token = cache.get(address)
+    if (!token) throw ERC721TokenDetailed_(chainId, erc721TokenContract, address, token)
+    return cache.get(address)!
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:43]:
```suggestion
    return token
```

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    } as ERC721TokenDetailed
+    cache.set(address, token_)
+    return
+}
+
+export function useAsyncERC721TokenDetailed(address: string, token?: Partial<ERC721TokenDetailed>) {
+    const chainId = useChainId()
+    const erc721TokenContract = useERC721TokenContract(address) as Contract
+    token = cache.get(address)
+    if (!token) throw ERC721TokenDetailed_(chainId, erc721TokenContract, address, token)
```
Comment (zhouhanseng) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:42]: I think refactor to suspense is not that easy, you need to consider retry and error handle. Please revert it if you are not familiar with suspense.

Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:42]: Actually, his implementation looks fine currently

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 16 above)
Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:20]: Don't mark it as resolved cause it is not resolved

**Review by Jack-Works — CHANGES_REQUESTED** [review]
Still have something we can make better

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 20]
```diff
@@ -31,14 +22,18 @@ async function ERC721TokenDetailed_(
         symbol: symbol ?? token?.symbol ?? '',
         baseURI: baseURI ?? token?.baseURI ?? '',
     } as ERC721TokenDetailed
-    cache.set(address, token_)
-    return
+
+    if (!cache.has(address)) cache.set(address, token_)
 }
 
-export function useAsyncERC721TokenDetailed(address: string, token?: Partial<ERC721TokenDetailed>) {
```
Comment (zhouhanseng) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts]: I don't think to integrate with suspense must rename useXXX to getXXX which cause react hooks should be moved out which breaks our code structure, check https://git.io/JtWVq

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+export function useAsyncERC721TokenDetailed(
+    address: string,
+    token?: Partial<ERC721TokenDetailed>,
+) {
+    const chainId = useChainId()
+    const erc721TokenContract = useERC721TokenContract(address)
+    const names = ['name', 'symbol', 'baseURI'] as (keyof ERC721['methods'])[]
+    const callDatas = new Array(3).fill([])
+    const [results, calls, _, callback] = useSingleContractMultipleData(erc721TokenContract, names, callDatas)
+    if (!cache.has(address)) throw getERC721TokenDetailed(chainId, address, results, callback, calls, token)
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:48]: I haven't written many suspenses before. It is a right way to mix suspense with hook?

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 22]
```diff
@@ -23,17 +27,28 @@ async function getERC721TokenDetailed_(
...
-    if (!cache.has(address)) throw getERC721TokenDetailed_(chainId, address, results, callback, calls, token)
-    return cache.get(address)!
+    const chainId = useChainId()
+    const erc721TokenContract = useERC721TokenContract(address)
+    const names = ['name', 'symbol', 'baseURI'] as (keyof ERC721['methods'])[]
+    const callDatas = new Array(3).fill([])
+    const [results, calls, _, callback] = useSingleContractMultipleData(erc721TokenContract, names, callDatas)
+    if (!cache.has(address)) throw getERC721TokenDetailed(chainId, address, results, callback, calls, token)
+    try {
+        return cache.get(address)!
```
Comment (zhouhanseng) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:50]: `cache.get(address)` won't cause any error, `try {} catch {}` is redundant.

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 23]
```diff
@@ -15,40 +15,30 @@ async function getERC721TokenDetailed(
...
+            type: EthereumTokenType.ERC721,
+            address: formatChecksumAddress(address),
+            chainId,
+            name: name ?? token?.name ?? '',
+            symbol: symbol ?? token?.symbol ?? '',
+            baseURI: baseURI ?? token?.baseURI ?? '',
+        } as ERC721TokenDetailed
+        cache.set(address, token_)
+    } catch {
+        cache.set(address, {} as ERC721TokenDetailed)
```
Comment (zhouhanseng) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:32]: @lelenei wait, this is not a the best way to cause an error.

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 21 above)
Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:42]: that's totally ok. you just don't violate the react hooks rule

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 20 above)
Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts]: Generally, it doesn't matter what it is named. But if you use hooks in it, it must be a hook too.

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 23 above)
Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:32]: so why not throw error directly?

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+    address: string,
+    results: any[],
+    callback: (calls_: any[]) => Promise<void>,
+    calls: any[],
+    token?: Partial<ERC721TokenDetailed>,
+) {
+    try {
+        await callback(calls)
+        // compose
+        const [name, symbol, baseURI] = results.map((x) => (x.error ? undefined : x.value))
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:21]: Hmm, I'm not familiar with this code flow but it seems a little strange to me. You're awaiting `callback(calls)` but don't use the return value of it, instead, you use `results` which comes from `useSingleContractMultipleData` hooks. When code recovered to run, `results` won't be updated (unless that hook is updating the original array directly). So I doubt is it working correctly?

Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:21]: maybe I should cc @guanbinrui or @zhouhanseng . `results` will be updated in the second pass IIUC but this code is expected to run once for each parameter combination (then it will set the cache, in the second pass, this code pass might not be run cause it already set the cache).

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 23 above)
Comment (zhouhanseng) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:32]: As I mentioned here: https://git.io/JtWyf

File: packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts (hunk 24 above)
Comment (zhouhanseng) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:21]: @lelenei

Comment (lelenei) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:21]: When the useERC721TokenDetailed returns, the content in value is replaced

Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:21]: let me check out this later

Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:21]: ok I actually got some knowledge from the react team about how to write suspense code correctly. got change this later

Comment (Jack-Works) [reply packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts:21]: ok I tried to rewrite this code but found it actually delegate data fetching to `useSingleContractMultipleData`. It therefore should not be an suspense resource. It is useful when you have an async function to get the result. In this case, which is not cause you don't have an async function.

BTW I'm not meaning this is wrong:

```ts
const id = useSomeHook()
return getSuspendedData(id)
```

This is totally correct. But in this case it's not.

```ts
const [results, calls, _, callback] = useSingleContractMultipleData(erc721TokenContract, names, callData)
if (!cache.has(address)) throw getERC721TokenDetailed(chainId, address, results, callback, calls, token)
```

cc @zhouhanseng @guanbinrui @lelenei

File: packages/maskbook/src/web3/hooks/useERC721TokenDetailed.ts [hunk 25]
```diff
@@ -6,19 +6,15 @@ import { useChainId } from './useChainState'
...
     // compose calls
-    const { names, callDatas } = useMemo(
-        () => ({
-            names: ['name', 'symbol', 'baseURI'] as 'name'[],
-            callDatas: new Array(3).fill([]),
-        }),
-        [],
-    )
+    const names = ['name', 'symbol', 'baseURI'] as (keyof ERC721['methods'])[]
+    const callDatas = new Array(3).fill([])
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/hooks/useERC721TokenDetailed.ts:17]: data cannot have plural.

File: packages/maskbook/src/plugins/COTM/UI/TokenAlbum.tsx [hunk 26]
```diff
@@ -27,26 +28,38 @@ const useStyles = makeStyles((theme) =>
...
+                            </section>
+                        ))}
+                    </div>
+                </div>
+            )}
+        </>
     )
 }
+function COTM_TOKEN_ADDRESS(COTM_TOKEN_ADDRESS: string) {
+    throw new Error('Function not implemented.')
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/COTM/UI/TokenAlbum.tsx:64]: ?

File: packages/maskbook/src/plugins/COTM/UI/TokenAlbum.tsx [hunk 27]
```diff
@@ -27,26 +28,38 @@ const useStyles = makeStyles((theme) =>
     }),
 )
 
-export interface TokenAlbumProps {}
+export interface TokenAlbumProps {
+    setCollectiblesLoading?: (loading: boolean) => void
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/COTM/UI/TokenAlbum.tsx:32]: please remove unused props

File: packages/maskbook/src/plugins/Election2020/UI/ElectionTokenAlbum.tsx [hunk 28]
```diff
@@ -28,28 +29,36 @@ const useStyles = makeStyles((theme) =>
     }),
 )
 
-export interface ElectionTokenAlbumProps {}
+export interface ElectionTokenAlbumProps {
+    setCollectiblesLoading?: (loading: boolean) => void
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Election2020/UI/ElectionTokenAlbum.tsx:33]: please remove unused props

**Review by Jack-Works — CHANGES_REQUESTED** [review]
